File: crates/vernier-core/examples/cache_speedup_val2017.rs
//! End-to-end measurement of [`BoundaryGtCache`] benefit on the COCO
//! val2017 perfect-segm workload.
//!
//! Models the training-loop validation pattern: same GT, fresh DT
//! each pass. The bench harness in `bench/` runs single-shot evaluates
//! per impl and so doesn't surface what the cache buys; this example
//! does two passes back-to-back in one process and reports both.
//!
//! Inputs (same conventions as the bench harness):
//! - `VERNIER_COCO_GT_PATH` → GT JSON (falls back to
//!   `~/.cache/vernier-bench/coco_val2017/instances_val2017.json`)
//! - `VERNIER_COCO_DT_SEGM_PATH` → DT JSON (falls back to
//!   `<repo>/.cache/coco-val2017/perfect_dt_segm.json`)
//!
//! Run:
//! ```sh
//! cargo run --release --example cache_speedup_val2017 -p vernier-core
//! ```

#![allow(clippy::unwrap_used, clippy::expect_used, clippy::print_stdout)]

use std::env;
use std::path::PathBuf;
use std::time::Instant;

use vernier_core::{
    evaluate_boundary, evaluate_boundary_cached, iou_thresholds, AreaRange, BoundaryGtCache,
    CocoDataset, CocoDetections, EvalDataset, EvaluateParams, ParityMode,
    BOUNDARY_DILATION_RATIO_DEFAULT,
};

fn gt_path() -> PathBuf {
    if let Ok(env) = env::var("VERNIER_COCO_GT_PATH") {
        return PathBuf::from(env);
    }
    let home = env::var("HOME").expect("HOME not set");
    PathBuf::from(home).join(".cache/vernier-bench/coco_val2017/instances_val2017.json")
}

fn dt_path() -> PathBuf {
    if let Ok(env) = env::var("VERNIER_COCO_DT_SEGM_PATH") {
        return PathBuf::from(env);
    }
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../.cache/coco-val2017/perfect_dt_segm.json")
}

fn time_ms<F: FnOnce() -> R, R>(f: F) -> (R, f64) {
    let t = Instant::now();
    let r = f();
    let ms = t.elapsed().as_secs_f64() * 1000.0;
    (r, ms)
}

fn main() {
    let gt_p = gt_path();
    let dt_p = dt_path();
    println!("GT: {}", gt_p.display());
    println!("DT: {}", dt_p.display());

    let (gt_bytes, load_gt_ms) = time_ms(|| std::fs::read(&gt_p).expect("read GT"));
    let (dt_bytes, load_dt_ms) = time_ms(|| std::fs::read(&dt_p).expect("read DT"));
    println!(
        "Loaded {:.1} MiB GT in {:.0} ms, {:.1} MiB DT in {:.0} ms",
        gt_bytes.len() as f64 / 1_048_576.0,
        load_gt_ms,
        dt_bytes.len() as f64 / 1_048_576.0,
        load_dt_ms,
    );

    let (gt, parse_gt_ms) =
        time_ms(|| CocoDataset::from_json_bytes(&gt_bytes).expect("parse GT"));
    let (dt, parse_dt_ms) =
        time_ms(|| CocoDetections::from_json_bytes(&dt_bytes).expect("parse DT"));
    println!("Parsed GT in {parse_gt_ms:.0} ms, DT in {parse_dt_ms:.0} ms");
    println!(
        "GT: {} images, {} categories; DT: {} detections",
        gt.images().len(),
        gt.categories().len(),
        dt.detections().len(),
    );

    let area = AreaRange::coco_default();
    let params = EvaluateParams {
        iou_thresholds: iou_thresholds(),
        area_ranges: &area,
        max_dets_per_image: 100,
        use_cats: true,
    };
    let ratio = BOUNDARY_DILATION_RATIO_DEFAULT;

    println!("\n=== uncached: two back-to-back evaluate_boundary calls ===");
    let (_, u1_ms) = time_ms(|| {
        evaluate_boundary(&gt, &dt, params, ParityMode::Strict, ratio).unwrap()
    });
    println!("call 1: {u1_ms:>9.0} ms");
    let (_, u2_ms) = time_ms(|| {
        evaluate_boundary(&gt, &dt, params, ParityMode::Strict, ratio).unwrap()
    });
    println!("call 2: {u2_ms:>9.0} ms");

    println!("\n=== cached: two back-to-back evaluate_boundary_cached calls (shared cache) ===");
    let cache = BoundaryGtCache::new();
    let (_, c1_ms) = time_ms(|| {
        evaluate_boundary_cached(&gt, &dt, params, ParityMode::Strict, ratio, &cache).unwrap()
    });
    println!(
        "call 1 (cold cache, populates): {c1_ms:>9.0} ms ({} entries)",
        cache.len()
    );
    let (_, c2_ms) = time_ms(|| {
        evaluate_boundary_cached(&gt, &dt, params, ParityMode::Strict, ratio, &cache).unwrap()
    });
    println!("call 2 (warm cache, hits):      {c2_ms:>9.0} ms");

    let speedup = u2_ms / c2_ms;
    let saved_ms = u2_ms - c2_ms;
    println!("\n=== summary ===");
    println!(
        "uncached pass-2 wall  : {u2_ms:>9.0} ms\n\
         cached   pass-2 wall  : {c2_ms:>9.0} ms\n\
         saved per warm call   : {saved_ms:>9.0} ms ({speedup:.2}× faster)"
    );
}

File: tools/cache_speedup_val2017.py
#!/usr/bin/env python3
"""End-to-end measurement of the parsed-once ``Dataset`` cache benefit
on the COCO val2017 perfect-segm workload (ADR-0020).

Models the training-loop validation pattern: same GT, fresh DT each
pass. The bench harness in ``bench/`` runs single-shot evaluates per
impl and so doesn't surface what the cache buys; this script does two
passes back-to-back in one process and reports both for both the
bytes-path and the ``Dataset``-path.

Mirrors the Rust example at
``crates/vernier-core/examples/cache_speedup_val2017.rs`` but exercises
the Python surface (``Evaluator(...).evaluate(Dataset, dt)``) so
regressions in the FFI/PyO3 wiring are caught here too.

Inputs (same conventions as the bench harness):
- ``VERNIER_COCO_GT_PATH`` -> GT JSON (falls back to
  ``~/.cache/vernier-bench/coco_val2017/instances_val2017.json``)
- ``VERNIER_COCO_DT_SEGM_PATH`` -> DT JSON (falls back to
  ``<repo>/.cache/coco-val2017/perfect_dt_segm.json``)

Run:
    uv run python tools/cache_speedup_val2017.py
"""

from __future__ import annotations

import os
import sys
import time
from collections.abc import Callable
from pathlib import Path

from vernier.instance import Boundary, Dataset, Evaluator


def _gt_path() -> Path:
    env = os.environ.get("VERNIER_COCO_GT_PATH")
    if env:
        return Path(env)
    return Path.home() / ".cache/vernier-bench/coco_val2017/instances_val2017.json"


def _dt_path() -> Path:
    env = os.environ.get("VERNIER_COCO_DT_SEGM_PATH")
    if env:
        return Path(env)
    repo_root = Path(__file__).resolve().parents[1]
    return repo_root / ".cache/coco-val2017/perfect_dt_segm.json"


def _timed(label: str, work: Callable[[], object]) -> float:
    t = time.perf_counter()
    work()
    ms = (time.perf_counter() - t) * 1000.0
    print(f"{label:<48} {ms:>9.0f} ms")
    return ms


def main() -> int:
    gt_p = _gt_path()
    dt_p = _dt_path()
    if not gt_p.exists():
        print(f"missing GT: {gt_p}", file=sys.stderr)
        return 1
    if not dt_p.exists():
        print(f"missing DT: {dt_p}", file=sys.stderr)
        return 1
    print(f"GT: {gt_p}")
    print(f"DT: {dt_p}")

    gt_bytes = gt_p.read_bytes()
    dt_bytes = dt_p.read_bytes()
    print(f"Loaded GT={len(gt_bytes) / 1_048_576:.1f} MiB, DT={len(dt_bytes) / 1_048_576:.1f} MiB")

    evaluator = Evaluator(iou=Boundary(), parity_mode="strict")

    print("\n=== bytes-path: two back-to-back evaluate calls (no cache) ===")
    _timed("bytes call 1", lambda: evaluator.evaluate(gt_bytes, dt_bytes))
    bytes2_ms = _timed("bytes call 2", lambda: evaluator.evaluate(gt_bytes, dt_bytes))

    print("\n=== Dataset-path: parse-once + warm-cache reuse ===")
    t = time.perf_counter()
    ds = Dataset.from_json(gt_bytes)
    print(f"{'Dataset.from_json':<48} {(time.perf_counter() - t) * 1000.0:>9.0f} ms")
    _timed("Dataset call 1 (cold cache, populates)", lambda: evaluator.evaluate(ds, dt_bytes))
    ds2_ms = _timed("Dataset call 2 (warm cache, hits)", lambda: evaluator.evaluate(ds, dt_bytes))

    speedup = bytes2_ms / ds2_ms if ds2_ms > 0 else float("inf")
    saved_ms = bytes2_ms - ds2_ms
    print("\n=== summary ===")
    print(f"bytes-path call 2  : {bytes2_ms:>9.0f} ms")
    print(f"Dataset call 2     : {ds2_ms:>9.0f} ms")
    print(f"saved per warm call: {saved_ms:>9.0f} ms ({speedup:.2f}x faster)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
